### PR TITLE
Prototype of FEM-PINNs

### PR DESCRIPTION
**Pull request by mtsokol**
Hi @pierremtb!

I'm a master degree student and recently I've been playing with PINNs source code and it's idea described in the paper. Amazing to see it rewritten to TF2.0! For me it was great way to learn `GradientTape` API.

I thought about introducing a small tweak to loss function as a way to merge FEM and NN for solving differential equations. I've spent on it last week I think I'm at a stage where I completed all I could think of (gradient is computed, error decreases) but still it does not solves it correctly and it's really slow.

I would be thankful if you could take a look at the implementation, I don't hold a lot of experience with Tensorflow so I could missed sth or the bare idea is faulty. Otherwise please let me know I'll close the PR.
(I've opened PR for convenience - will close it).

---
# Idea

With standard FEM method we would define our base functions and search for the coefficients that convey problem defined by differential equation and initial and border conditions. So `u` that we search for would be approximated by combination of coefficients and base functions.
We would do it by creating and solving some system of equations.

In PINNs we train a NeuralNet that will behave as a direct solution for our `u` by placing equation in the loss function.

Here the idea is a combination of both. We want to train a NeuralNet that for given element (e.g. (i,j) for 2D problem) will return its coefficient. So the solution will be described by basis functions but its coefficients are acquired from NN. Here I decided to use BSplines as basis functions.

So the loss function won't change:

But our approximation of `u`:

And with our NN:

# Code

In this PR I've tried to complete it as on paper it looked feasible. I've created a new Layer that computes `u` approximation (together with comments). I've used BSpline basis from `tensorflow_graphics` library. 

**For better and faster explanation I've prepared short notebook** how that bspline implementation  works together with plots (it's the same implementation so I used it for testing):
https://github.com/mtsokol/PINNs-TF2.0/blob/FEM-PINNs-TF2.0/1d-burgers/BSplines.ipynb

The rest is all the same. It runs with `python3 1d-burgers/inf_cont_burgers.py`

# Problems

I've made it in 40 lines of code and spent a bit of time on that but unfortunately I didn't make it work as expected. I've got to major issues:

1. The error is decreasing but unfortunately all coefficients are basically pulled down close to zero (I thought about scaling output but learning still just pulls it down) so there is no sinus recreated as initial condition (but with those spline basis it's possible). What could I do to check it whether it's just impossible with such loss function or is it implementation error or maybe wrong RL, weights initialization, output scaling, etc.?
2. It's really slow. For 1000 collocation points Adam is slow but LBFGS takes literally forever. In PR I'm using 100 colloc points: Adam -> 0.5 sec per 10 iter., LBFGS 35 sec. (!). and final sampling is also scaled down.

So here I think it's due to the way gradient is computed. In standard version we just multiply in input by layers, here for input we first compute coefficients from NN (easy) and then for each element we compute spline evaluation so there is few multiplication and also cond  with also makes this gradient harder to compute. 

I've tried to introduce `vectorized_map` or `iter=10` in `map_fn` but both fail with different error that I couldn't solve.

How can I confirm whether it's such inefficient complexity of loss function or whether implementation issue?

----

Sorry once again for such long post. 

What do you think about this idea? Does it make sense? Or maybe it's inefficient approach when it comes to complexity? How can I work on problems described?
If you have any advises regarding implementation I would be thankful as I still learning Tensorflow.

### CODE REVIEW

**Comment by pierremtb** [comment]
Hi Mateusz, thanks for reaching out! Unfortunately, I'm incredibly busy ATM, so is it fine if I take a deeper look at this later this week?

**Comment by mtsokol** [comment]
@pierremtb Sure! Thank you once again for any help. 

(p.s. in case for known problem with installing latest `tensorflow_graphics` dependency on mac here's solution https://github.com/AcademySoftwareFoundation/openexr/issues/567#issuecomment-590144773)

**Comment by mtsokol** [comment]
@pierremtb Hi! 
Sorry for interrupting once again. Did you have time to look at it?

**Comment by pierremtb** [comment]
Hey @mtsokol, I'm very sorry it must have fell through. Will try to dive in today or tomorrow!

File: 1d-burgers/inf_cont_burgers.py [hunk 1]
```diff
@@ -28,16 +28,20 @@
...
     # Collocation points size, where we’ll check for f = 0
-    hp["N_f"] = 10000
+    # [mtsokol] for bigger values slows down drastically
+    hp["N_f"] = 100
     # DeepNN topology (2-sized input [x t], 8 hidden layer of 20-width, 1-sized output [u]
-    hp["layers"] = [2, 20, 20, 20, 20, 20, 20, 20, 20, 1]
+    # [mtsokol] 7x7=49 for one_hot_encoding, more about it in neuralnetwork.py
+    # [mtsokol] IMPORTANT! when modifying input size remember to modify BSplineLayer(num_elem=) param. (Should be done in one place only. Will fix it)
+    # [mtsokol] num_elem = sqrt(layers[0]) - 2 (only for one_hot_encoding)
+    hp["layers"] = [49, 20, 20, 20, 20, 20, 20, 20, 20, 1]
```
Comment (pierremtb) [inline 1d-burgers/inf_cont_burgers.py:37]: Have you tried to lower the number of layers? It has a big impact on computational cost. Maybe you could try say 2 layers of 50 and a larger N_f

File: 1d-burgers/inf_cont_burgers.py [hunk 2]
```diff
@@ -28,16 +28,20 @@
...
     # Setting up the TF SGD-based optimizer (set tf_epochs=0 to cancel it)
     hp["tf_epochs"] = 100
     hp["tf_lr"] = 0.03
     hp["tf_b1"] = 0.9
     hp["tf_eps"] = None
     # Setting up the quasi-newton LBGFS optimizer (set nt_epochs=0 to cancel it)
-    hp["nt_epochs"] = 200
+    hp["nt_epochs"] = 100
```
Comment (pierremtb) [inline 1d-burgers/inf_cont_burgers.py]: TBH I never really was impressed with LBGFS, Adam does the job well more often than not (esp for experimentation)

File: 1d-burgers/inf_cont_burgers.py [hunk 3]
```diff
@@ -106,16 +110,19 @@ def predict(self, X_star):
     X_u_train, u_train, X_f, ub, lb = prep_data(
         path, hp["N_u"], hp["N_f"], noise=0.0)
 
+# [mtsokol] for bigger values slows down drastically
+X_star = X_star[::100, :]
+u_star = u_star[::100, :]
```
Comment (pierremtb) [inline 1d-burgers/inf_cont_burgers.py:115]: What's going on here?

File: utils/neuralnetwork.py [hunk 4]
```diff
@@ -1,8 +1,53 @@
...
+            row = np.zeros((input_shape,))
+            row[i] = 1
+            one_hot_encoded.append(row)
+        self.cofs = tf.convert_to_tensor(one_hot_encoded, dtype=tf.float32) # cofs ->> coefficients' indices
+
+    def call(self, input):
+        output = self.cofs # first we need to ask our NN for coefficients for each 2D BSpline...
+        for layer in self.layers:
+            output = layer(output)
+        u = tf.reshape(output, [self.matrix_size,self.matrix_size]) * 3.0 # ...and reshape it. IMPORTANT! I think we need to multiply it to scale bc otherwise all coefficients will be almost 0 and with BSplines values result ends up as 0
```
Comment (pierremtb) [inline utils/neuralnetwork.py]: Would it make sense to work with log or something to deal with small numbers?

**Review by pierremtb — COMMENTED** [review]
Hey, finally I get a chance to dive in a bit. Thanks for the PR it's a wonderful idea. Haven't looked deeply in the TF layers stuff, but I left a few simple comments

File: 1d-burgers/inf_cont_burgers.py (hunk 1 above)
Comment (mtsokol) [reply 1d-burgers/inf_cont_burgers.py:37]: Thank you for your comments!
So I tried with only two layers, different input magnitude but it still just pull coefficients close to zero. 
I've checked bspline logic once again for errors (same as in notebook which I linked) and it seems fine so I'm not sure if such procedure can converge to desired solution. 

Should I commit those layer changes?

File: 1d-burgers/inf_cont_burgers.py (hunk 2 above)
Comment (mtsokol) [reply 1d-burgers/inf_cont_burgers.py]: Sure, here it's more handy to test with lbgfs turned off. Should I commit this?

File: 1d-burgers/inf_cont_burgers.py (hunk 3 above)
Comment (mtsokol) [reply 1d-burgers/inf_cont_burgers.py:115]: With full `X_star` it takes a lot of time to compute final error so I taken every hundredth to make it faster.

File: utils/neuralnetwork.py (hunk 4 above)
Comment (mtsokol) [reply utils/neuralnetwork.py]: So with the `tanh` activation the output (coefficients) can be from -1 to 1. The spline basis functions are in range 0 to 1:
 

So after multiplying basis and coefs I think it might be impossible to easily achieve greater values as final output like >8.0 here in `sin` initial condition.

That's the reason why I scaled output by multiplying by 3 (also tried different values), it gave bigger initial error but again was pulled close to zero as before.

I don't understand `play icon ce` - could you explain it to me?

File: 1d-burgers/inf_cont_burgers.py (hunk 1 above)
Comment (pierremtb) [reply 1d-burgers/inf_cont_burgers.py:37]: No don't commit. But was the performance better?

File: 1d-burgers/inf_cont_burgers.py (hunk 2 above)
Comment (pierremtb) [reply 1d-burgers/inf_cont_burgers.py]: It kinda over complicates the code as so yeah I would say so.

File: utils/neuralnetwork.py (hunk 4 above)
Comment (pierremtb) [reply utils/neuralnetwork.py]: Sorry about this, autocorrect got me. I corrected it to "deal with small numbers"

File: 1d-burgers/inf_cont_burgers.py (hunk 1 above)
Comment (mtsokol) [reply 1d-burgers/inf_cont_burgers.py:37]: Slightly better but without significant improvement.
